File: scan.php
<?php
// Patterns of suspicious code
$suspicious_patterns = [
    '/eval\(base64_decode\(/',   // Common obfuscation technique
    '/eval\(/',                 // Generic eval usage
    '/base64_decode\(/',        // Base64 decoding
    '/shell_exec\(/',           // Executes shell commands
    '/passthru\(/',             // Executes shell commands
    '/exec\(/',                 // Executes shell commands
    '/system\(/',               // Executes system commands
    '/preg_replace\(.+\/e/',    // Deprecated /e modifier in preg_replace
    '/gzinflate\(/',            // Inflates compressed strings
];

// Function to scan a single file
function scanFile($filePath, $patterns) {
    $results = [];
    $lines = file($filePath, FILE_IGNORE_NEW_LINES | FILE_SKIP_EMPTY_LINES);
    
    if ($lines === false) {
        return ["Error reading file"];
    }

    foreach ($lines as $lineNumber => $lineContent) {
        foreach ($patterns as $pattern) {
            if (preg_match($pattern, $lineContent)) {
                $results[] = [
                    'line' => $lineNumber + 1, // Line number (starting from 1)
                    'pattern' => $pattern,
                    'content' => trim($lineContent)
                ];
            }
        }
    }

    return $results;
}

// Function to recursively scan all files in a directory
function scanDirectory($directory, $patterns) {
    $results = [];
    $files = new RecursiveIteratorIterator(new RecursiveDirectoryIterator($directory));

    foreach ($files as $file) {
        if ($file->isFile()) {
            $filePath = $file->getRealPath();
            $scanResults = scanFile($filePath, $patterns);
            if (!empty($scanResults)) {
                $results[$filePath] = $scanResults;
            }
        }
    }

    return $results;
}

// Define the directory to scan
$targetDirectory = __DIR__; // Current directory (public_html)

// Start scanning
echo "Scanning directory: $targetDirectory\n";
$scanResults = scanDirectory($targetDirectory, $suspicious_patterns);

if (!empty($scanResults)) {
    echo "\nSuspicious files detected:\n";
    foreach ($scanResults as $file => $issues) {
        echo "File: $file\n";
        foreach ($issues as $issue) {
            echo "  Line: {$issue['line']} | Pattern: {$issue['pattern']} | Content: {$issue['content']}\n";
        }
        echo "\n";
    }
} else {
    echo "\nNo suspicious files detected.\n";
}
?>

File: scan.py
import os
import re

# Path to the folder to scan
TARGET_DIR = "../public_html"

# Patterns of suspicious code
SUSPICIOUS_PATTERNS = [
    r"eval\(base64_decode\(",   # Common obfuscation technique
    r"eval\(",                 # Generic eval usage
    r"base64_decode\(",        # Base64 decoding
    r"shell_exec\(",           # Executes shell commands
    r"passthru\(",             # Executes shell commands
    r"exec\(",                 # Executes shell commands
    r"system\(",               # Executes system commands
    r"preg_replace\(.*/e",     # Deprecated /e modifier in preg_replace
    r"gzinflate\(",            # Inflates compressed strings (often used in obfuscation)
]

def scan_file(file_path):
    """Scan a single file for suspicious patterns."""
    results = []
    try:
        with open(file_path, 'r', encoding='utf-8', errors='ignore') as f:
            for line_num, line in enumerate(f, start=1):
                for pattern in SUSPICIOUS_PATTERNS:
                    if re.search(pattern, line):
                        results.append((line_num, pattern))
    except Exception as e:
        return False, f"Error reading file: {e}"
    return results

def scan_directory(directory):
    """Recursively scan all files in a directory."""
    results = []
    for root, _, files in os.walk(directory):
        for file in files:
            file_path = os.path.join(root, file)
            suspicious_matches = scan_file(file_path)
            if suspicious_matches:
                for line_num, pattern in suspicious_matches:
                    results.append((file_path, line_num, pattern))
    return results

def main():
    print(f"Scanning directory: {TARGET_DIR}")
    results = scan_directory(TARGET_DIR)
    if results:
        print("\nSuspicious files detected:")
        for file_path, line_num, pattern in results:
            print(f"File: {file_path} | Line: {line_num} | Pattern: {pattern}")
    else:
        print("\nNo suspicious files detected.")

if __name__ == "__main__":
    main()
